### ISSUE TITLE
BUG: upper cap < 3.12

### ISSUE THREAD

**Issue by anna-charlotte** [scipy/scipy#17957]
### Describe your issue.

We use `poetry` for dependency management and encounter the following problem when trying to add scipy:
Due to the upper cap on python 3.12 we can not add scipy via poetry, since we do no have such an upper cap.

Would it be possible to remove this upper cap or change it to 3.13?

### Reproducing Code Example

```python
-
```

### Error message

```shell
(venv_new) >>>>charlotte@MacBook ~/Desktop/project % poetry add scipy     

Using version ^1.10.0 for scipy

Updating dependencies
Resolving dependencies... (0.2s)

The current project's Python requirement (>=3.8) is not compatible with some of the required packages Python requirement:
  - scipy requires Python <3.12,>=3.8, so it will not be satisfied for Python >=3.12
```

### SciPy/NumPy/Python version information

1.10.0

**Comment by tupui** [comment]
Hi @anna-charlotte, thank you for reporting.

This is intentional as said in our release notes: SciPy 1.10 does not support Python 3.12. 

We are very careful about our pins and since we have a complex build toolchain, new versions of Python usually mean something will break and Python 3.12 is still in alpha.

We only start to support new Python versions when they enter the RC phase because during the beta phase we have time to fix things.

**Comment by tupui** [comment]
Just to backup a bit my claim, I am adding a 3.12 job to our CI and NumPy is failing to build https://github.com/scipy/scipy/actions/runs/4144130190/jobs/7166811829
i.e. things are not ready.

**Comment by sisp** [comment]
I think capping the Python version is _really_ a **bad** idea and considered bad practice:

* https://github.com/pypa/packaging.python.org/pull/850
* https://iscinumpy.dev/post/bound-version-constraints/#pinning-the-python-version-is-special
* https://discuss.python.org/t/requires-python-upper-limits/12663

**Comment by ilayn** [comment]
If you want to go down the rabbit hole where we discussed this many many times already, and we know what you mean. Here is the canonical place that includes many SciPy maintainers discussing this with Python core members. 

https://discuss.python.org/t/requires-python-upper-limits/12663

What is typically missed out in these discussions is that we have a multiple version span of NumPy and Python versions (and many other details like compiler/dependence problems Cython/pybind11 etc). 

Upper capping is done for planned obsolescence and expectation management instead of letting the code fail in numerous cryptic ways.

People tend to downplay the breakage happening between different Python versions probably they are used to work with Python only packages but we have exceeding amount of ABI issues and that forces us to be strict. Hence you might not be convinced but read through all the discussions and linked issues and it will give you at least the reasoning we have. You can disagree but we are doing it this way to save a ton of maintainer work and we are not given any other solution so far.

**Comment by tupui** [comment]
> It looks like [SciPy v1.11.1](https://github.com/scipy/scipy/blob/v1.11.1/setup.py?rgh-link-date=2023-07-20T09%3A22%3A25Z#L454-L458) caps the Python version to <3.13 (i.e. Python 3.12 is included), but wheel files are only available for Python <3.12 and I can't find CI checks for Python 3.12. 🤔

Yes, as I said above the current versions of SciPy don't install on Python 3.12, hence no wheel. The cap on <3.13 is because we will make a patch release as soon as 3.12 is out.

We are confident we will have this figured out by then. We have an opened PR to test on 3.12 and someone is looking at fixing the issues we have here.

In any case, we won't release a wheel until Python 3.12 is officially released.

> Poetry for dependency management, the (really strict) Python version capping gets propagated

Yes, this is an issue with how Poetry think Python should deal with dependencies, not taking into account the fact that sometimes maintainers do know what they are doing 😉

> A change in NumPy may break SciPy, but likely not all of it but only a small fraction of the API. 

We are very strict here. SciPy is only released if the test suite is passing. We literally have millions of users and you would be surprised to see how SciPy is used. People rely on this stability and we are not going to change our position here.

Some projects are using our nightlies to test their library. SciPy is breaking things like scikit-image, scikit-learn and networkX all the time. And it's not always a very used function which breaks.

> To restore functionality, SciPy would need a patch release.

Making a patch release is not trivial at all. To the end user, they just see a wheel, but getting there requires many manual steps on our side. Due to the complexity of the project and the time needed to make a release, we only do patch releases in between major versions. Meaning as soon as 1.12 would be out, we would stop making patch releases for 1.11.

**Comment by andyfaff** [comment]
> I can't find CI checks for Python 3.12

That's now being done.

**Comment by andyfaff** [comment]
Nightly wheels for cp312 are now available. Please see https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy. An official wheel won't be released until after Python 3.12 has an official release.
